src/scifem/mesh.py

```diff
@@ -103,3 +103,64 @@ def transfer_meshtags_to_submesh(
         entity_tag._cpp_object, submesh.topology._cpp_object, vertex_to_parent, cell_to_parent
     )
     return dolfinx.mesh.MeshTags(cpp_tag), sub_to_parent_entity_map
+
+
+def reverse_mark_entities(
+    entity_map: dolfinx.common.IndexMap, entities: npt.NDArray[np.int32]
+) -> npt.NDArray[np.int32]:
+    """Communicate entities marked on a single process to all processes that ghosts or
+    owns this entity.
+
+    Args:
+        entity_map: Index-map describing entity ownership
+        entities: Local indices of entities to communicate
+    Returns:
+        Local indices marked on any process sharing this entity
+    """
+    comm_vec = dolfinx.la.vector(entity_map, dtype=np.int32)
+    comm_vec.array[:] = 0
+    comm_vec.array[entities] = 1
+    comm_vec.scatter_reverse(dolfinx.la.InsertMode.add)
+    comm_vec.scatter_forward()
+    return np.flatnonzero(comm_vec.array).astype(np.int32)
+
+
+def extract_submesh(
+    mesh: dolfinx.mesh.Mesh, entity_tag: dolfinx.mesh.MeshTags, tags: tuple[int, ...]
+) -> tuple[
+    dolfinx.mesh.Mesh,
+    npt.NDArray[np.int32],
+    npt.NDArray[np.int32],
+    npt.NDArray[np.int32],
+    dolfinx.mesh.MeshTags,
+]:
```

Comment:
Might consider to return a named tuple here. I think in general that if one return more than three things one should group it into a named tuple (but this is a personal preference).